lib/tree.js

```diff
@@ -12,40 +12,115 @@ class Tree {
+        if (!current) {
+            return newNode;
+        } else {
+            if (current.key > newNode.key) {
+                current.left = this.addHelper(current.left, newNode);
+            } else if (current.key <= newNode.key) {
+                current.right = this.addHelper(current.right, newNode);
+            }
+        }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+        return current;
+    }
+
+    // Time Complexity: O(log n)
+    // Space Complexity: O(log n)
     find(key) {
-        throw new Error("This method hasn't been implemented yet!");
+        if (!this.root) {
+            return null;
+        } else {
+            return this.findHelper(this.root, key);
+        }
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    findHelper(current, key) {
+        if (!current) {
+            return false;
+        } else { 
+            if (current.key == key) {
+                return current.value;
+            } else if (current.key > key) {
+                return (this.findHelper(current.left, key));            
+            } else if (current.key < key) {
+                return (this.findHelper(current.right, key));            
+            }
+        }
+    }
+
+    // Time Complexity: O(n)
+    // Space Complexity: O(log n)
     inorder() {
-        throw new Error("This method hasn't been implemented yet!");
+        return this.inorderHelper(this.root, []);
+    }
+    inorderHelper(current, values) {
+        if (!!current) {
+            this.inorderHelper(current.left, values);
+            values.push({key: current.key, value: current.value});
+            this.inorderHelper(current.right, values);
+        }
+        return values;
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    // Time Complexity: O(n)
+    // Space Complexity: O(log n)
     preorder() {
-        throw new Error("This method hasn't been implemented yet!");
+        return this.preorderHelper(this.root, []);
+    }
+    preorderHelper(current, values) {
+        if (!!current) {
+            values.push({key: current.key, value: current.value});
+            this.preorderHelper(current.left, values);
+            this.preorderHelper(current.right, values);
+        }
+        return values;
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    // Time Complexity: O(n)
+    // Space Complexity: O(log n)
     postorder() {
-        throw new Error("This method hasn't been implemented yet!");
+        return this.postorderHelper(this.root, []);
+    }
+    postorderHelper(current, values) {
+        if (!!current) {
+            this.postorderHelper(current.left, values);
+            this.postorderHelper(current.right, values);
+            values.push({key: current.key, value: current.value});
+        }
+        return values;
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    // Time Complexity: O(log n)
+    // Space Complexity: O(log n)
     height() {
```

Comment:
👍 The time complexity is O(n) since you are visiting each node and the space complexity is O(log n) if the tree is unbalanced and O(log n) if it is balanced.